File: 2020/day1/c++/part2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

typedef vector<int>::iterator NumIt;

int prod_sum(NumIt first, NumIt last, int sum, int n)
{
    int p = -1;
    for (auto i = first; i != last-1; i++) {
        int sought = sum - *i;
        if (n == 2) {
            auto it = lower_bound(i+1, last, sought);
            if (it != last && *it == sought) {
                p = *i * sought;
                break;
            }
        } else {
            int prod = prod_sum(i+1, last, sought, n-1);
            if (prod < 0)
                continue;
            p = *i * prod;
            break;
        }
    }
    return p;
}

int main()
{
    string s;
    vector<int> numbers;
    ifstream input_file;

    input_file.open("../input.txt", ios::in);
    if (!input_file) {
        cerr << "cannot open file" << endl;
        exit(1);
    }
    while (getline(input_file, s))
        numbers.push_back(stoi(s));
    input_file.close();

    if (numbers.empty())
        return 0;

    sort(numbers.begin(), numbers.end());

    cout << prod_sum(numbers.begin(), numbers.end(), 2020, 3) << endl;

    return 0;
}

File: 2020/day1/java/Part2.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;

class Part2
{
    private static List<Integer> readFile()
    {
        List<Integer> numbers = new ArrayList<Integer>();
        try {
            File inputFile = new File("../input.txt");
            Scanner inputScanner = new Scanner(inputFile);
            while (inputScanner.hasNextLine()) {
                String line = inputScanner.nextLine();
                int x = Integer.parseInt(line);
                numbers.add(x);
            }
            inputScanner.close();
        } catch (FileNotFoundException e) {
            System.out.println("Cannot open file");
        }
        return numbers;
    }

    public static void main(String[] args)
    {
        List<Integer> numbers = readFile();
        Collections.sort(numbers);

        for (int i = 0 ; i < numbers.size(); i++) {
            for (int j = i+1 ; j < numbers.size(); j++) {
                int x = numbers.get(i);
                int y = numbers.get(j);
                Integer sought = 2020 - x - y;
                int idx = numbers.indexOf(sought);
                if (idx > j && idx != -1) {
                    System.out.println(x*y*numbers.get(idx));
                }
            }
        }
    }
}
